### PR TITLE
fix: add lazy compilation to handle HMR failure for new initial chunk…

### LINKED ISSUE #18548: hmr might fail if there are new initial chunks
# Bug report

**What is the current behavior?**

HMR might fail if one initial chunk occur at second compilation.

The reason:

If one chunk is not initial chunk in the first compilation, but is initial chunk in the second compilation, when apply hot updates, there will be some imports to the initial chunk, but at this moment, the initial chunks are not loaded, so webpack will go fetching that chunk, and treats it as async chunk, so if async chunk path is different from initial chunk path, it cannot find that chunk

**If the current behavior is a bug, please provide the steps to reproduce.**

repro repo:

https://github.com/JSerFeng/webpack-lazy-compilation-issue/tree/hmr

run

```bash
pnpm i
pnpm run dev
```

open the browser

modify `src/initial.js`

```diff
try {
-  require("core-js/package.json");
+  require("core-js");
} catch (_err) {}

module.hot.accept();

module.exports = 1;
```

**What is the expected behavior?**

**Other relevant information:**
webpack version:  5.92.1

### PR DESCRIPTION
**Pull request by Vansh5632**
fixes #18548 
What kind of change does this PR introduce?
This PR introduces a bugfix to handle HMR (Hot Module Replacement) failures caused by newly added initial chunks. Lazy compilation has been added to resolve this issue.

Did you add tests for your changes?
Yes, tests have been added to verify the implementation of lazy compilation and ensure HMR functions correctly when new initial chunks are introduced.

Does this PR introduce a breaking change?
No, this PR does not introduce any breaking changes. It enhances the existing functionality without impacting current workflows.

What needs to be documented once your changes are merged?

The addition of lazy compilation for handling HMR with new initial chunks.
Steps or configurations (if any) required to enable lazy compilation.
Potential performance improvements or edge cases addressed by this fix.

### CODE REVIEW

File: hot/lazy-compilation-node.js [hunk 1]
```diff
@@ -35,16 +47,53 @@ exports.keepAlive = function (options) {
...
+	}
+
+	/**
+	 * Handle connection errors.
+	 * @param {Error} err Error object
 	 */
 	function errorHandler(err) {
 		err.message =
 			"Problem communicating active modules to the server: " + err.message;
+		console.error(err.message);
```
Comment (alexander-akait) [inline hot/lazy-compilation-node.js:78]: Do we need it?

File: hot/lazy-compilation-node.js [hunk 2]
```diff
@@ -35,16 +47,53 @@ exports.keepAlive = function (options) {
...
 		onError(err);
+
+		// Force reload for critical errors
+		if (module.hot) {
+			console.error("Critical error detected. Forcing reload...");
+			throw new Error("Critical error detected. Forcing reload..."); // Throw an error to trigger a restart
```
Comment (alexander-akait) [inline hot/lazy-compilation-node.js:84]: I don't think we need it

File: hot/lazy-compilation-node.js (hunk 1 above)
Comment (Vansh5632) [reply hot/lazy-compilation-node.js:78]: yes it's also in the original code

File: hot/lazy-compilation-node.js (hunk 2 above)
Comment (Vansh5632) [reply hot/lazy-compilation-node.js:84]: if not i might check this out and resolve it

**Review by invisiblepancake — APPROVED** [review]

**Review by invisiblepancake — APPROVED** [review]

**Review by alexander-akait — COMMENTED** [review]
Looks good, could you add it a test case too?

**Review by alexander-akait — CHANGES_REQUESTED** [review]
@Vansh5632 I don't think this fix is correct - try https://github.com/webpack/webpack/issues/18548, also we need a valid test case, i.e. using node and web target, also we have `test/hotCases/lazy-compilation` directory for tests

**Comment by Vansh5632** [comment]
> @Vansh5632 I don't think this fix is correct - try #18548, also we need a valid test case, i.e. using node and web target, also we have `test/hotCases/lazy-compilation` directory for tests

can you elaborate it a little more

**Comment by Vansh5632** [comment]
unassign

**Review by invisiblepancake — COMMENTED** [review]
=)

**Pull request closed without merging**